### PR TITLE
NN-3075 Filter by transaction type

### PR DESCRIPTION
**Pull request by JamieDraperUK**
- Filter by transaction type
- Make date range optional

### CODE REVIEW

File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/repository/OffenderTransactionHistoryRepository.java [hunk 1]
```diff
@@ -11,9 +11,9 @@
 
 public interface OffenderTransactionHistoryRepository extends CrudRepository<OffenderTransactionHistory, Pk> {
 
-    @Query("select h from OffenderTransactionHistory h where h.offender.nomsId = ?1 and h.accountType = ?2 and (h.entryDate between ?3 and ?4)")
-    List<OffenderTransactionHistory> findForGivenAccountType(String offenderNo, String accountType, LocalDate fromDate, LocalDate toDate);
-
-    @Query("select h from OffenderTransactionHistory h where h.offender.nomsId = ?1 and (h.entryDate between ?2 and ?3)")
-    List<OffenderTransactionHistory> findForAllAccountTypes(String offenderNo, LocalDate fromDat, LocalDate toDate);
+    @Query("select h from OffenderTransactionHistory h where h.offender.nomsId = :offenderNo and " +
+        " (:accountType is null or h.accountType = :accountType) and " +
```
Comment (JamieDraperUK) [inline src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/repository/OffenderTransactionHistoryRepository.java:15]: Move to a single repository method with optional parameters due to the filter combinations getting out of hand.

File: src/main/java/uk/gov/justice/hmpps/prison/api/resource/OffenderResource.java [hunk 2]
```diff
@@ -93,9 +93,9 @@
     private final OffenderTransactionHistoryService offenderTransactionHistoryService;
```
Comment (JamieDraperUK) [inline src/main/java/uk/gov/justice/hmpps/prison/api/resource/OffenderResource.java:93]: Formatting changes.

File: src/main/java/uk/gov/justice/hmpps/prison/api/resource/OffenderResource.java [hunk 3]
```diff
@@ -369,22 +369,28 @@ private Status lookupStatusOrDefaultToAll(final String status) {
     }
 
     @ApiResponses(value = {
```
Comment (JamieDraperUK) [inline src/main/java/uk/gov/justice/hmpps/prison/api/resource/OffenderResource.java:371]: Endpoint changes.

File: src/test/java/uk/gov/justice/hmpps/prison/service/OffenderTransactionHistoryServiceTest.java [hunk 4]
```diff
@@ -42,174 +43,112 @@ public void setUp() {
...
             final Optional<LocalDate> toDateOpl = Optional.of(LocalDate.now());
-            service.getTransactionHistory(OFFENDER_NO, accountCode, fromDateOpl, toDateOpl);
+            service.getTransactionHistory(OFFENDER_NO, accountCode, fromDateOpl, toDateOpl, null);
         });
 
         assertEquals("accountCode optional can't be null", exception.getMessage());
     }
 
     @Test
-    public void When_getTransactionHistory_And_FromDateIsNull_Then_ThrowException() {
```
Comment (JamieDraperUK) [inline src/test/java/uk/gov/justice/hmpps/prison/service/OffenderTransactionHistoryServiceTest.java:109]: No longer needed

File: src/test/java/uk/gov/justice/hmpps/prison/service/OffenderTransactionHistoryServiceTest.java [hunk 5]
```diff
@@ -42,174 +43,112 @@ public void setUp() {
...
             final Optional<LocalDate> toDateOpl = Optional.of(LocalDate.now().plusDays(2));
-            service.getTransactionHistory(OFFENDER_NO, accountCode, fromDateOpl, toDateOpl);
+            service.getTransactionHistory(OFFENDER_NO, accountCode, fromDateOpl, toDateOpl, null);
         });
 
         assertEquals("toDate can't be in the future", exception.getMessage());
     }
 
     @Test
-    public void When_getTransactionHistory_And_DatesDefaultToNow_Then_CallRepositoryWithoutAccountCode() {
```
Comment (JamieDraperUK) [inline src/test/java/uk/gov/justice/hmpps/prison/service/OffenderTransactionHistoryServiceTest.java:174]: Removed as no longer needed.

File: src/test/java/uk/gov/justice/hmpps/prison/service/OffenderTransactionHistoryServiceTest.java [hunk 6]
```diff
@@ -361,44 +298,14 @@ public void When_getTransactionHistory_ShouldBe_SortedByEntryDateDescending() {
     }
 
     @Test
-    public void When_getTransactionHistory_ForAccountCode_Maps_Correctly() {
```
Comment (JamieDraperUK) [inline src/test/java/uk/gov/justice/hmpps/prison/service/OffenderTransactionHistoryServiceTest.java:364]: Removed as no longer needed.

File: src/test/java/uk/gov/justice/hmpps/prison/service/OffenderTransactionHistoryServiceTest.java [hunk 7]
```diff
@@ -420,43 +327,16 @@ public void When_getTransactionHistory_Maps_Correctly() {
     }
 
     @Test
-    public void When_getTransactionHistoryRelatedTransactionDetails_Correctly() {
```
Comment (JamieDraperUK) [inline src/test/java/uk/gov/justice/hmpps/prison/service/OffenderTransactionHistoryServiceTest.java:423]: Removed as no longer needed.

File: src/main/java/uk/gov/justice/hmpps/prison/service/OffenderTransactionHistoryService.java [hunk 8]
```diff
@@ -29,62 +29,59 @@
...
+                          final Optional<LocalDate> toDate) {
 
         checkNotNull(offenderNo, "offenderNo can't be null");
         checkNotNull(accountCodeOpl, "accountCode optional can't be null");
-        checkNotNull(fromDateOpl, "fromDate optional can't be null");
-        checkNotNull(toDateOpl, "toDate optional can't be null");
+        checkNotNull(fromDate, "fromDate optional can't be null");
+        checkNotNull(toDate, "toDate optional can't be null");
+
+        if (fromDate.isPresent() && toDate.isPresent()) checkDateRange(fromDate.get(), toDate.get());
```
Comment (mikehalmamoj) [inline src/main/java/uk/gov/justice/hmpps/prison/service/OffenderTransactionHistoryService.java:76]: I think I'm missing something here, but won't the checkNotNull throw an exception before you get to this line, so no need to check `isPresent`?  Also, not really sure why these parameters are nullable in the API only to then validate them?

Comment (JamieDraperUK) [reply src/main/java/uk/gov/justice/hmpps/prison/service/OffenderTransactionHistoryService.java:76]: You're absolutely right! Good spot! I'll fix now.

Comment (mikehalmamoj) [reply src/main/java/uk/gov/justice/hmpps/prison/service/OffenderTransactionHistoryService.java:76]: Oh sorry, Optional vs null - two different things. :facepalm:

**Review by mikehalmamoj — APPROVED** [review]

**Pull request merged**